MPQ / MC exercise editing

nathanstitt opened the pull request:
@jivewise and I collaborated on this by using sub-pr's and merging them in when ready.

![screen shot 2016-04-07 at 5 58 42 pm](https://cloud.githubusercontent.com/assets/79566/14369405/6304ac22-fcea-11e5-83e5-029b74ef73a0.png)

nathanstitt committed:
First commit towards new exercise editing flow

nathanstitt committed:
remove stray merge marker

nathanstitt committed:
hook up new exercise action

nathanstitt committed:
Style questions

nathanstitt committed:
Connect save/publish actions, tags component

nathanstitt committed:
Display attachements on assets tab

jivewise committed:
Merge pull request #51 from openstax/mpq-editor-ns

Display attachements on assets tab

nathanstitt committed:
Some tag styles

nathanstitt committed:
Move tags into directory, render book

nathanstitt committed:
use `book:` prefix for book tag

nathanstitt committed:
Load exercise id if it's in url

nathanstitt committed:
LO/Books error handling and adding multiple types

nathanstitt committed:
Set tab content to be white with border

nathanstitt committed:
Show all tags on preview

nathanstitt committed:
remove margin to prevent content jumps

nathanstitt committed:
Make padding apply to all panes

jivewise committed:
Merge pull request #53 from openstax/mpq-editor-ns

tag inputs with validation

nathanstitt committed:
remove dead code

nathanstitt committed:
comment for method use

nathanstitt committed:
Question Type tag

nathanstitt committed:
Use indexOf to reject tags that start with prefix

nathanstitt committed:
remove console.log :(

nathanstitt committed:
Add react-widgets library

For use by tags that are multi-select

nathanstitt committed:
Reset all prefixed tags with array

nathanstitt committed:
filter-type tag

nathanstitt committed:
format-type tag

nathanstitt committed:
Add "Requires Choice" checkbox

nathanstitt committed:
remove formats tag, it's a per-question field

nathanstitt committed:
use prefix constant

nathanstitt committed:
CNX tags, new MultiInput with validation

nathanstitt committed:
DOK, Bloom, & Time fields

nathanstitt committed:
Error monitoring and display

nathanstitt committed:
return tags sorted from store vs sort in component

nathanstitt committed:
Exercise ID

nathanstitt committed:
Remove dead tag methods

nathanstitt committed:
Boostrap up form controls so appearance matches

nathanstitt committed:
remove more tag fixed methods

nathanstitt committed:
Hide delete controls when only a single tag is left

nathanstitt committed:
Remove exid tag when deleting it's book

nathanstitt committed:
exid tags can't be deleted by themselves

nathanstitt committed:
Add missed delete method

nathanstitt committed:
When adding a blank ,check for equality not start

nathanstitt committed:
No dupes, only add tag if it doesn't exist.

jivewise committed:
adding in exercise stimulus styles and responsive embed

jivewise committed:
remove question link style

jivewise committed:
moving exercise-wrapper code into app and handling new exercises better

jivewise committed:
allow adding in new questions in exercises, and allow removal of questions in multipart

jivewise committed:
updating stimulus and stem order, and adding in remove question link

jivewise committed:
updating stores to use @_get instead of @_local

jivewise committed:
Merge pull request #54 from openstax/mpq-editor-ns

Remainder of tag types

jivewise committed:
Merge pull request #55 from openstax/mpq-editor-errors

Display error modal with details when an error occurs

jivewise committed:
moving error modal code into app, and removing exercise-wrapper

jivewise committed:
remove empty arrays from templates

jivewise committed:
hide attachments if exercise is new, there's no uid, attachments need a uid

jivewise committed:
hide publish button if exercise is new, and use async buttons to indicate progress

nathanstitt committed:
kill commented code left-over from prev component

nathanstitt committed:
Include link to internal support form on error modal

nathanstitt committed:
Merge pull request #56 from openstax/mpq-editor-jw

Exercises: Multipart Questions

nathanstitt committed:
Render 500 error message as HTML

nathanstitt committed:
line length blah, blah

jivewise committed:
Merge pull request #57 from openstax/mpq-editor-ns

Include link to support form in error modal

nathanstitt committed:
rename "Preserve Answer Order" to "Order Matters"

nathanstitt committed:
Rename "Free Response" to "Requires viewing ..."

jivewise committed:
renaming button to be clearer

jivewise committed:
renaming button to be clearer

jivewise committed:
renaming button to be clearer

jivewise committed:
removing stimulus and adding in ordering questions ui

jivewise committed:
removing stimulus and adding in ordering questions ui

jivewise committed:
removing stimulus and adding in ordering questions ui

jivewise committed:
question move arrow styles

jivewise committed:
question move arrow styles

jivewise committed:
question move arrow styles

jivewise committed:
capitalize intro tab

jivewise committed:
capitalize intro tab

jivewise committed:
Merge branch 'mpq-editor-jw' of https://github.com/openstax/exercises-js into mpq-editor-jw

nathanstitt committed:
Display warning when converting a MPQ to MC

nathanstitt committed:
Remove stray console.log

jivewise committed:
adding preview back in

jivewise committed:
adding preview back in

jivewise committed:
Merge pull request #58 from openstax/mpq-editor-ns

Warning when converting MPQ to MC

jivewise committed:
Merge branch 'mpq-editor-jw' of https://github.com/openstax/exercises-js into mpq-editor-jw

nathanstitt committed:
Merge pull request #59 from openstax/mpq-editor-jw

Question Reordering

nathanstitt committed:
show feedback and answers on preview

nathanstitt committed:
pull in latest

nathanstitt committed:
Copy collaborator_solutions to just solutions for preview

jivewise committed:
Merge pull request #60 from openstax/mpq-editor-ns

Preview update

nathanstitt committed:
Use updated exercse preview that uses collaborator_solutions

nathanstitt committed:
Switch between vertical / horizontal layout

nathanstitt committed:
kill tabs and re-indent

nathanstitt committed:
Merge pull request #61 from openstax/mpq-editor-ns

Switch between vertical / horizontal layout

nathanstitt committed:
Update react-components to preview release

nathanstitt committed:
Only apply rules to direct descendant

.controls is to generic, it's used by other controls on the page like
the tag headings

nathanstitt committed:
Rename question labels

nathanstitt committed:
Bind to store vs calling .sync and add QuestionFormat

nathanstitt committed:
Question format-type control

nathanstitt committed:
adjust for responsive screens

nathanstitt committed:
use questions format control with requires choices input

nathanstitt committed:
Allow either one or two digits

nathanstitt committed:
Use forceUpdate instead of empty setState

nathanstitt committed:
a bit of margin to the top of tags

nathanstitt committed:
remove unused files

nathanstitt committed:
requires-context tag

nathanstitt committed:
Add requires-context tag, and formatting cleanups

nathanstitt committed:
open-ended isn't free-response

nathanstitt committed:
Move logic into store for requires choices

jivewise committed:
Merge pull request #62 from openstax/mpq-editor-formats

Questions formats control and field renames

nathanstitt committed:
Include some missing requires

nathanstitt committed:
Move toggle to left so it doesn't overflow when split

nathanstitt committed:
Copy in current test setups

nathanstitt committed:
Upgrade react-components to latest

No changes other than some warnings silenced

nathanstitt committed:
A few minimal specs for exercises

nathanstitt committed:
Merge pull request #63 from openstax/mpq-editor-specs

Specs for exercises

nathanstitt committed:
Fix checkbox default value

nathanstitt committed:
Formats is null on a new exercise

nathanstitt committed:
Re-add needed sync() calls

They're needed to update the preview

nathanstitt committed:
remove old specs

nathanstitt committed:
Require spec files so Travis runs them

nathanstitt committed:
Upgrade to react-components version w/out lodash

nathanstitt committed:
Remove stray console.log

nathanstitt committed:
Disable publishing twice

nathanstitt committed:
Move book select into it's own control

This way it can be shared with LO

nathanstitt committed:
LO has the book id prefixed

nathanstitt committed:
Only allow choosing valid books for LO tag

nathanstitt committed:
Merge pull request #64 from openstax/mpq-editor-ns

LO tag with book ID

philschatz commented on package.json:
```diff
@@ -7,33 +7,33 @@
     "font-awesome": "4.3.0",
     "htmlparser2": "3.8.3",
     "jquery": "2.1.4",
+    "katex": "0.5.1",
```
Are we still using katex?

philschatz commented on package.json:
```diff
@@ -7,33 +7,33 @@
-    "openstax-react-components": "openstax/react-components#53ed6ed38b6752523e85104c3f7ef7c5c36debc8",
+    "openstax-react-components": "openstax/react-components#bc7d2a1bd2a48cd991920f250dad8a3aadcd8ce0",
     "pluralize": "1.1.2",
+    "quill-with-math": "philschatz/quill-with-math#master",
```
We're still using quill too?

nathanstitt replied:
I guess so?? Those aren't new, the changes are the result of the sorting the dependencies, when I installed `react-widgets` (for it's multi-select).

philschatz replied:
Ah, could you remove them? I'm _pretty sure_ they're no longer being used

philschatz commented on resources/styles/exercises.less:
```diff
@@ -22,96 +38,76 @@ body {
-  .tags select {
-    width: 100%;
-    margin-bottom: 1rem;
+  > * {
```
Is there a more-specific selector that could be used?

philschatz commented on resources/styles/question.less:
```diff
@@ -0,0 +1,69 @@
+.question {
+	.move-question {
```
Can you fix up the whitespace?

philschatz commented on resources/styles/question.less:
```diff
@@ -0,0 +1,69 @@
+		}
+	}
+  textarea {
+    display: block;
```
Is this not already `display: block`?

nathanstitt replied:
Nope, actually `<textarea>` defaults to `inline-block`.  With that said though, I don't see why it needs to be set to `block`, it still renders fine when I remove the rule.  This file is extracted from when all styles where just one big file, so I'm not sure of the reasoning for setting it to `block`

philschatz commented on src/components/app.cjsx:
```diff
@@ -1,35 +1,140 @@
+    if @props.exerciseId is 'new'
+      @addNew()
+    else if @props.exerciseId
+      this.refs.exerciseId.getDOMNode().value = @props.exerciseId
```
Could this be done using `<BS.Input defaultValue={@props.exerciseId}/>`?

nathanstitt replied:
great catch! I think it's probably fairly dangerous to rely on refs being set in a `didMount` 

Followup edit: Actually It's safe in a `didMount` since that's called after the initial render.  Still a bad practice though

philschatz commented on src/components/app.cjsx:
```diff
@@ -1,35 +1,140 @@
+  renderEditingBody: ->
+    return null unless @state.exerciseId
+    if ExerciseStore.isLoading()
+      <div>Loading...</div>
```
can you add classes similar to `<Loadable>` so selenium tests would work (even though we don't have them for exercises yet)?

philschatz commented on src/components/app.cjsx:
```diff
@@ -1,35 +1,140 @@
+                </AsyncButton>
+              }
+              <form className="navbar-form navbar-right">
+                <BS.Input type="checkbox" label=" Split Vertically"
```
Ah, why did you opt for a checkbox instead of responsive (via Bootstrap's grid)?

nathanstitt replied:
Lack of Time ;)  Actually I've now decided that the way to go is just split vertically all the time but with a max-width on the preview.

philschatz commented on src/components/error-modal.cjsx:
```diff
@@ -0,0 +1,76 @@
+# coffeelint: disable=max_line_length
```
Is this Error modal the same as in `tutor-js`?

nathanstitt replied:
It's similar, but has a few differences.  This one attempts to render the HTML that exercises BE sends back on 500's, and has a different support link.  I considered extracting the Tutor error modal, but decided they were too different and relied to much on binding to a flux store.

philschatz commented on src/components/exercise.cjsx:
```diff
@@ -2,166 +2,141 @@
-        >
-        Publish
-      </AsyncButton>
+    question = @state.tab.split('-')[1]
```
hmm, is there no better way to find the question index than parsing `"question-42"`?

nathanstitt replied:
I looked into this a bit and the answer I've came to is "not very easily".  The tab controls `onChange` method only receives a single argument, the tab's `eventKey`.  That makes it difficult to use a `data-question-index` or other attribute for determining the current tab.

philschatz commented on src/components/question-format-type.cjsx:
```diff
@@ -0,0 +1,60 @@
+
+  updateFormat: (ev) ->
+    formats =  QuestionStore.get(@props.questionId).formats or []
+    if ev.target.checked
```
_Minor:_ Could you use `@refs.theCheckbox.checked()` instead of the event directly?

nathanstitt replied:
Hm, not very easily.  There's not a single checkbox, the `updateFormat` method is called by all the checkboxes `onChange` (one for each format type). 

I suppose I could use `_.partial` to create custom callback methods with the type bound to the handler, but it seems like just using the event's values is a bit lighter weight.

philschatz commented on src/components/question-format-type.cjsx:
```diff
@@ -0,0 +1,60 @@
+    isChoiceRequired = QuestionStore.isChoiceRequired(@props.questionId)
+
+    <div className="format-type">
+      {for id, name of TYPES
```
_Aside:_ I like the readability of the nested for loops :smile:

philschatz commented on src/components/question.cjsx:
```diff
@@ -57,50 +68,47 @@ module.exports = React.createClass
+
+      {if removeQuestion
+        <div>
+          {if canMoveRight
```
Odd, I would have expected `canMoveLeft` to appear first in the DOM (but I don't have a screenshot to look at for reference)

nathanstitt replied:
Hm, yes this is a bit weird.  It's because they're all `float: right` so the order is reversed.  Let me see if we can remove some of the floats

Followup edit:  I made the containing div float right and then the controls appear in the correct order in the DOM

philschatz commented on src/components/tags/book-selection.cjsx:
```diff
@@ -0,0 +1,39 @@
+React = require 'react'
+_ = require 'underscore'
+
+BOOKS =
```
Would it be easier to update constants like these if they were in a file higher up (like in `src/constants.json`)?

nathanstitt replied:
yeah, not really sure on that.  I expect the process of someone who's attempting to update the book lists to be: find tags/book component.  Discover it renders the shared book-selection component and then update the list.  Given your suggestion it'd be an additional bit of detective work to find the src/constants.json file.   

Ideally, I'd say that the BE would have an `/admin` page like Tutor that this could be set in, but we're not quite there yet.

philschatz commented on src/components/tags/cnx-mod.cjsx:
```diff
@@ -0,0 +1,28 @@
+    exerciseId: React.PropTypes.string.isRequired
+
+  validateInput: (value) ->
+    'Must match CNX feature' unless value.match(
```
Should this be `'Must match CNX module ID (without the version number)'`?

philschatz commented on src/components/tags/cnx-mod.cjsx:
```diff
@@ -0,0 +1,28 @@
+      prefix='cnx-module'
+      cleanInput={@cleanInput}
+      validateInput={@validateInput}
+      placeholder='1d1fd537-77fb-4eac-8a8a-60bbaa747b6d'
```
Could this be `Format: #########-####-4###-####-############` so users are not confused as to why a UUID is in the input but gray?

philschatz commented on src/components/tags/index.cjsx:
```diff
@@ -0,0 +1,40 @@
+        <CnxFeature      {...@props} />
+        <Dok             {...@props} />
+        <Blooms          {...@props} />
+        <Time            {...@props} />
```
Nice! (listing them all off like that here)

philschatz commented on package.json:
```diff
@@ -7,33 +7,33 @@
     "test": "gulp test"
   },
   "dependencies": {
-    "font-awesome": "4.3.0",
```
From the code it seems like you're still using font-awesome, no?

nathanstitt replied:
it moved down to line 16 when items were sorted ;)

philschatz commented on test/all-tests.coffee:
```diff
@@ -1,4 +1,2 @@
-it 'should pass a stub test', ->
-  expect(true).to.be.true
+testsContext = require.context("./", true, /\.spec\.(cjsx|coffee)$/)
+testsContext.keys().forEach(testsContext)
```
Slick!

philschatz commented:
# :muscle:

Nice work! Even the hairy code dealing with tags is really readable!

nathanstitt committed:
Remove vertical split option in favor of max-width

nathanstitt committed:
Remove dependency on katex and quill packages

nathanstitt committed:
re-indent and remove tab characters

nathanstitt committed:
Remove block style from textarea

nathanstitt committed:
Check exerciseId prop in loadExercise method

nathanstitt committed:
Set loading classes

nathanstitt committed:
Float entire div to right and vertically align controls

nathanstitt committed:
Use ## for placeholder and better wording on error message

nathanstitt committed:
Remove old_components directory

nathanstitt commented:
@philschatz I think I've addressed all comments, ready for another round of review.
@jivewise I really hope the fixes do not conflict with your work on tests.  I didn't make any substantial changes to source at least.
